app/uk/gov/hmrc/merchandiseinbaggage/controllers/GoodsTypeQuantityController.scala

```diff
@@ -50,11 +52,19 @@ class GoodsTypeQuantityController @Inject()(
       .fold(
         formWithErrors =>
           Future.successful(BadRequest(view(formWithErrors, idx, request.declarationJourney.declarationType, backButtonUrl(idx)))),
-        categoryQuantityOfGoods =>
-          persistAndRedirect(
-            request.goodsEntry.copy(maybeCategoryQuantityOfGoods = Some(categoryQuantityOfGoods)),
-            idx,
-            routes.GoodsVatRateController.onPageLoad(idx))
+        categoryQuantityOfGoods => {
+          val updatedGoodsEntry = request.goodsEntry match {
+            case entry: ImportGoodsEntry => entry.copy(maybeCategoryQuantityOfGoods = Some(categoryQuantityOfGoods))
+            case entry: ExportGoodsEntry => entry.copy(maybeCategoryQuantityOfGoods = Some(categoryQuantityOfGoods))
+          }
```

Comment:
i know why you ended up with same logic in both bases, but may be having a common model as before might help reduce the amount of code like above if we structure  the `originOrDestinationOfGoods` as 
```Either[ProducedInTheEu, Country]``` ?